src/threads-version/big_int64.cpp

```
@@ -0,0 +1,767 @@
+  uint64_t high, low, carry;
+
+  for (int i = 0; i < n; i++) {
+    carry = 0;
+    for (int j = 0; j < m; j++) {
+      mul2Limbs64(a.limbs[i], b.limbs[j], high, low);
+      adder3_64(low, result[i + j], carry, result[i + j], carry);
+      carry += high;
+    }
+    // Handle any remaining carry after the inner loop
+    result[i + m] = carry;
+  }
+
+  BigInt64 res;
+  res.limbs = std::move(result);
+  res.removeLeadingZeros(); // Remove leading zeros
+  res.isNegative = a.isNegative != b.isNegative;
+  return res;
+}
+
+void BigInt64::exportTo(uint8_t *out, size_t outLen, CreateModes mode) const {
+  if (outLen < bytesCount())
+    throw std::runtime_error("Not enough space in output buffer");
+  if (mode == CreateModes::LITTLE_ENDIANESS)
+    for (int i = 0; i < limbsCount(); i++)
+      for (int j = 0; j < BYTES_IN_LIMB; j++)
+        out[outLen - 1 - i * BYTES_IN_LIMB - (7 - j)] =
+            static_cast<uint8_t>(limbs[i] >> (j * BITS_IN_BYTE));
+  else
+    for (int i = 0; i < limbsCount(); i++)
+      for (int j = 0; j < BYTES_IN_LIMB; j++)
+        out[outLen - 1 - i * BYTES_IN_LIMB - j] =
+            static_cast<uint8_t>(limbs[i] >> (j * BITS_IN_BYTE));
+}
+
+int BigInt64::bitsCount() const {
+
+  if (isZero())
+    return 0;
+
+  // Count bits in the most significant limb manually
+  uint64_t mostSignificantLimb = limbs.back();
+  int inMsb = ::bitsCount(mostSignificantLimb);
+
+  return inMsb + (limbsCount() - 1) * BITS_IN_LIMB;
+}
+void BigInt64::thisEqualsbBSubthis(const BigInt64 &b) {
+  BigInt64 copyB(b);
+  copyB.subtractUnsigned(*this); // copyB-=this
+  *this = std::move(copyB);
+}
+void BigInt64::removeLeadingZeros() {
+  while (limbs.size() > 1 && limbs.back() == 0) {
+    limbs.pop_back();
+  }
+}
+void BigInt64::insertZroes(int n) { limbs.insert(limbs.begin(), n, 0); }
+void BigInt64::generateNLimbsRandom(int limbsCnt) {
+  if (limbsCnt < 1)
+    throw std::invalid_argument("limbsCnt less than 1");
```

Should return an error code?

---

wrapped with try catch